### PR TITLE
[Jormun] Sort administration region by level

### PR DESCRIPTION
**Pull request by nonifier**
We have recently broken the interface by changing the order of which administration regions get returned, like in :
```
{
  "administrative_regions": [
    {  
      "id": "admin:osm:relation:1773820",
      "level": 10,
      "name": "Wazemmes",
      [...]
    },
    {
      "id": "admin:fr:59350",
      "level": 8,
      "name": "Lille (59000-59800)",
      [...]
    },
    {
      "id": "admin:fr:95354",
      "level": 9,
      "name": "Lille",
      [...]
    }
]
```

In order not to break the previous implementation, we need to sort the admin regions by `level`.
This is achieve by implementing the `SortedGenericSerializer` interface.

:warning:  I broke it down by commits to make it easier to read :)

todo:
 - [x] Run Artemis - https://ci.navitia.io/job/custom_Artemis/27/ :ok_hand:

### CODE REVIEW

File: source/jormungandr/jormungandr/interfaces/v1/serializer/base.py [hunk 1]
```diff
@@ -210,6 +210,26 @@ class PbGenericSerializer(PbNestedSerializer):
...
+    def to_value(self, obj):
+        fields = self._compiled_fields
+        serialize = self._serialize
+
+        if self.many == False:
+            return serialize(obj, fields)
+
+        serialized_objs = (serialize(o, fields) for o in obj)
+        return sorted(serialized_objs, key=self.sort_key)
+
```
Comment (xlqian) [inline source/jormungandr/jormungandr/interfaces/v1/serializer/base.py:236]: you can add a abstracted method to this class
```python
    @abstractmethod
    def sort_key(self):
        raise NotImplemented
```

File: source/type/type_interfaces.h [hunk 2]
```diff
@@ -100,6 +100,9 @@ enum class OdtLevel_e { scheduled = 0, with_stops = 1, zonal = 2, all = 3 };
 struct Nameable {
     std::string name;
     bool visible = true;
+
+    Nameable() = default;
+    Nameable(const std::string name) : name(name) {}
```
Comment (xlqian) [inline source/type/type_interfaces.h]: const std::string& or std::move

File: source/jormungandr/tests/places_tests.py [hunk 3]
```diff
@@ -296,3 +296,16 @@ def test_places_has_no_distance(self):
...
+class TestAdministrativesRegions(AbstractTestFixture):
+    def test_administrative_regions_are_ordered_by_level(self):
+        """
+        In kraken's data, administrative regions aren't sorted.
+        At serialization time, we should order them by level ascending
+        """
+        response = self.query_region("stop_areas/A")
+        admins = response['stop_areas'][0]['administrative_regions']
+        levels = [admin['level'] for admin in admins]
+        assert levels[0] < levels[1] < levels[2]
```
Comment (benoit-bst) [inline source/jormungandr/tests/places_tests.py:311]: :+1:

**Review by benoit-bst — APPROVED** [review]

File: source/jormungandr/jormungandr/interfaces/v1/serializer/base.py [hunk 4]
```diff
@@ -210,6 +211,31 @@ class PbGenericSerializer(PbNestedSerializer):
     name = jsonschema.Field(schema_type=str, display_none=True, description='Name of the object')
 
 
+@six.add_metaclass(abc.ABCMeta)
+class SortedGenericSerializer(object):
```
Comment (xlqian) [inline source/jormungandr/jormungandr/interfaces/v1/serializer/base.py]:
```python
class SortedGenericSerializer(serpy.Serializer):
```

File: source/jormungandr/jormungandr/interfaces/v1/serializer/base.py [hunk 5]
```diff
@@ -32,6 +32,7 @@
 import serpy
 import six
 import operator
+import abc
```
Comment (xlqian) [inline source/jormungandr/jormungandr/interfaces/v1/serializer/base.py:35]: this is useless, no?(almost!! :D )

**Review by xlqian — APPROVED** [review]

File: source/jormungandr/jormungandr/interfaces/v1/serializer/base.py (hunk 5 above)
Comment (nonifier) [reply source/jormungandr/jormungandr/interfaces/v1/serializer/base.py:35]: I need it for :
```python
    @abc.abstractmethod
    def sort_key(self):
        raise NotImplementedError
```
:smiley:

Comment (xlqian) [reply source/jormungandr/jormungandr/interfaces/v1/serializer/base.py:35]: oooooppppps :smile:

File: source/jormungandr/jormungandr/interfaces/v1/serializer/pt.py [hunk 6]
```diff
@@ -309,13 +314,16 @@ class CarParkSerializer(PbNestedSerializer):
     total_places = jsonschema.IntField()
 
 
-class AdminSerializer(PbGenericSerializer):
+class AdminSerializer(SortedGenericSerializer, PbGenericSerializer):
```
Comment (pbench) [inline source/jormungandr/jormungandr/interfaces/v1/serializer/pt.py:317]: Don't you end up with some Diamond inheritance problem ? 
Both SortedGenericSerializer and PbGenericSerializer inherits from serpy.Serializer

**Review by pbench — APPROVED** [review]

File: source/jormungandr/jormungandr/interfaces/v1/serializer/pt.py (hunk 6 above)
Comment (xlqian) [reply source/jormungandr/jormungandr/interfaces/v1/serializer/pt.py:317]: let me see

Comment (xlqian) [reply source/jormungandr/jormungandr/interfaces/v1/serializer/pt.py:317]: https://www.programiz.com/python-programming/multiple-inheritance

take a look at "MRO" :)

**Pull request merged**